### PR TITLE
Allow `unread` of `AbstractVector{UInt8}`

### LINKED ISSUE #210: Allow `unread` of `AbstractVector{UInt8}`
Currently `unread` only works with `Vector{UInt8}`

https://github.com/JuliaIO/TranscodingStreams.jl/blob/8fdabeda88199cd4298d9cba597e006b1af7655a/src/stream.jl#L445

It should be changed to allow any `AbstractVector{UInt8}`

Ref: https://github.com/JuliaIO/TranscodingStreams.jl/pull/209#discussion_r1590356676

### PR DESCRIPTION
**Pull request by nhz2**
Fixes #210

### CODE REVIEW

File: src/buffer.jl [hunk 1]
```diff
@@ -202,7 +202,9 @@ function copydata!(data::Ptr{UInt8}, buf::Buffer, nbytes::Integer)
 end
 
 # Insert data to the current buffer.
-function insertdata!(buf::Buffer, data::Ptr{UInt8}, nbytes::Integer)
+# `data` must not alias `buf`
+function insertdata!(buf::Buffer, data::Union{AbstractArray{UInt8}, Memory})
```
Comment (mkitti) [inline src/buffer.jl:206]: This change looks breaking. Is this public API?

File: src/buffer.jl [hunk 2]
```diff
@@ -213,7 +215,9 @@ function insertdata!(buf::Buffer, data::Ptr{UInt8}, nbytes::Integer)
     end
     datasize = buf.marginpos - datapos
     copyto!(buf.data, datapos + nbytes, buf.data, datapos, datasize)
-    GC.@preserve buf unsafe_copyto!(bufferptr(buf), data, nbytes)
+    for i in 0:nbytes-1
```
Comment (mkitti) [inline src/buffer.jl:218]: Could you use `copyto!`?

**Review by mkitti — COMMENTED** [review]
We may need to preserve the public API or deprecate that particular method.

You could also consider `Base.require_one_based_indexing` if there is any question about the applicability to `AbstractVector`.

File: src/buffer.jl (hunk 1 above)
Comment (nhz2) [reply src/buffer.jl:206]: No, `insertdata!` is internal and not part of the docs.

https://juliahub.com/ui/Search?q=insertdata!&type=code

File: src/buffer.jl (hunk 2 above)
Comment (nhz2) [reply src/buffer.jl:218]: `copyto!` isn't defined for `TranscodingStreams.Memory`, but it does have `getindex`

Comment (nhz2) [reply src/buffer.jl:218]: Also `copyto!` seems like it has special cases for handling aliasing. https://github.com/JuliaLang/julia/issues/50900
This isn't needed here.

**Review by mkitti — APPROVED** [review]

**Pull request merged**